Sources/NIOSOCKS/Channel Handlers/SOCKSClientHandler.swift

```diff
@@ -31,11 +33,20 @@ public final class SOCKSClientHandler: ChannelDuplexHandler {
     private var inboundBuffer: ByteBuffer?
     
     private var bufferedWrites: MarkedCircularBuffer<(NIOAny, EventLoopPromise<Void>?)> = .init(initialCapacity: 8)
+    private let connectionEstablishedPromise: EventLoopPromise<Void>?
     
     /// Creates a new `SOCKSClientHandler` that connects to a server
     /// and instructs the server to connect to `targetAddress`.
     /// - parameter targetAddress: The desired end point - note that only IPv4, IPv6, and FQDNs are supported.
-    public init(targetAddress: SOCKSAddress) {
+    public convenience init(targetAddress: SOCKSAddress) {
+        self.init(targetAddress: targetAddress, connectionEstablishedPromise: nil)
+    }
+    
+    /// Creates a new `SOCKSClientHandler` that connects to a server
+    /// and instructs the server to connect to `targetAddress`.
+    /// - parameter targetAddress: The desired end point - note that only IPv4, IPv6, and FQDNs are supported.
+    /// - parameter connectionEstablishedPromise: An optional promise that is succeeded once the SOCKS connection is established.
+    public init(targetAddress: SOCKSAddress, connectionEstablishedPromise: EventLoopPromise<Void>?) {
```

Comment:
Having thought about this for even longer, I'm not sure if this is the best solution (given that this shall become public API). The problem at hand is that using this initializer becomes very hard in `.channelInitializer`, since they might be invoked multiple times. Of course we could make a `var connectionEstablishedFuture: EventLoopFuture<Void>?` public, but using this API would be quite bad.